Adding multigammaln ref and fix arange

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #85333
* __->__ #85153

Partially based on https://github.com/pytorch/pytorch/pull/83662.

I'll help land this one, as Rob does not work in the PyTorch project
anymore

I removed the data-dependent check for the args, as data dependencies
are bad for many reasons (and it was failing when the input has NaNs).

It also registers arange as a decomposition, and fixes the naming of its
args.

lezcano committed:
Land "Adding multigammaln ref"

Partially based on https://github.com/pytorch/pytorch/pull/83662.

I'll help land this one, as Rob does not work in the PyTorch project
anymore

I removed the data-dependent check for the args, as data dependencies
are bad for many reasons (and it was failing when the input has NaNs).

It also registers arange as a decomposition, and fixes the naming of its
args.

[ghstack-poisoned]

Lezcano commented:
This one will also need to be coordinated with https://github.com/pytorch/pytorch/pull/84957 (@nkaretnikov)

lezcano committed:
Update on "Land "Adding multigammaln ref""


Partially based on https://github.com/pytorch/pytorch/pull/83662.

I'll help land this one, as Rob does not work in the PyTorch project
anymore

I removed the data-dependent check for the args, as data dependencies
are bad for many reasons (and it was failing when the input has NaNs).

It also registers arange as a decomposition, and fixes the naming of its
args.

[ghstack-poisoned]

Lezcano commented:
@ngimel @mruberry this one should be ready

mruberry commented on test/test_unary_ufuncs.py:
```diff
@@ -624,16 +624,6 @@ def test_frexp_assert_raises(self, device):
                 ):
                     torch.frexp(input, out=(mantissa, exponent))
 
-    def test_mvlgamma_argcheck(self, device):
```
The docs will need to be updated; currently they say an error will be thrown in this case:

https://pytorch.org/docs/master/special.html

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -3630,7 +3628,8 @@ def arange(
         )
-    elif a is not None and b is None:
+    else:
+        assert a is not None and b is None
```
Update this to use `utils.check()`

mruberry replied:
(Also this should add a comment as to what this clause is for)

mruberry replied:
The if/else pattern here seems a little awkward to me vs. just populating the `start` value if not provided and then going through one code path

mruberry replied:
Wait... does this implement the logic correctly? 

```
torch.arange(end=5)
```

Is valid -- how would that work here?

Lezcano replied:
Yeah, this logic felt weird, and it is indeed incorrect. I'll fix it on Monday.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5575,22 +5575,20 @@ def skips_mvlgamma(skip_redundant=False):
 # Class `MvlGammaInfo` already contains the basic information related to the operator,
```
I think the pR needs to update this comment if it's switching the class for a factory function

lezcano committed:
Update on "Land "Adding multigammaln ref""


Partially based on https://github.com/pytorch/pytorch/pull/83662.

I'll help land this one, as Rob does not work in the PyTorch project
anymore

I removed the data-dependent check for the args, as data dependencies
are bad for many reasons (and it was failing when the input has NaNs).

It also registers arange as a decomposition, and fixes the naming of its
args.

[ghstack-poisoned]

Lezcano commented on torch/_decomp/decompositions.py:
```diff
@@ -1579,7 +1579,7 @@ def index_add_(
             utils.is_weakly_lesser_type(type(alpha), python_type),
             lambda: f"alpha argument of type {type(alpha)} cannot be safely cast to type {python_type}!",
         )
-        tensor = torch._prims.mul(tensor, alpha)
```
Fixes https://github.com/pytorch/pytorch/pull/85002#discussion_r973632113

Lezcano commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -740,7 +740,10 @@ def to_float(start, end, step):
+            # Pass end as positional arg
             yield SampleInput(end, kwargs={"dtype": dtype, "device": device})
+            # (Similar to) calling torch.arange(end=3)
+            yield SampleInput(0, kwargs={"end": end, "dtype": dtype, "device": device})
```
This one we cannot test properly really. This is a limitation of OpInfos.

mruberry replied:
A limitation of how we express SampleInputs, you mean? Because we assume the SampleInput has at least one positional argument?

Lezcano replied:
exactly

Lezcano replied:
In retrospect, I think it wasn't a good design to single out the first arg to `SampleInput`. Taking the positional args as `*args` and the rest kw-args may have done the trick and would be more extensible.

lezcano committed:
Update on "Adding multigammaln ref and fix arange"


Partially based on https://github.com/pytorch/pytorch/pull/83662.

I'll help land this one, as Rob does not work in the PyTorch project
anymore

I removed the data-dependent check for the args, as data dependencies
are bad for many reasons (and it was failing when the input has NaNs).

It also registers arange as a decomposition, and fixes the naming of its
args.

[ghstack-poisoned]

lezcano committed:
Update on "Adding multigammaln ref and fix arange"


Partially based on https://github.com/pytorch/pytorch/pull/83662.

I'll help land this one, as Rob does not work in the PyTorch project
anymore

I removed the data-dependent check for the args, as data dependencies
are bad for many reasons (and it was failing when the input has NaNs).

It also registers arange as a decomposition, and fixes the naming of its
args.

[ghstack-poisoned]

lezcano committed:
Update on "Adding multigammaln ref and fix arange"


Partially based on https://github.com/pytorch/pytorch/pull/83662.

I'll help land this one, as Rob does not work in the PyTorch project
anymore

I removed the data-dependent check for the args, as data dependencies
are bad for many reasons (and it was failing when the input has NaNs).

It also registers arange as a decomposition, and fixes the naming of its
args.

[ghstack-poisoned]

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -3752,31 +3726,22 @@ def arange(
-        )
-    else:
-        raise AssertionError()
+    assert not pin_memory
```
`assert` -> `utils.check`

Lezcano replied:
I just wrote it as an assert as it's more of a hard fail as it's not implemented. I could use a `utils.check` with a `NotImplmentedError`, but I think shorter is better here.

Lezcano replied:
Addressed here https://github.com/pytorch/pytorch/pull/85333

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -3752,31 +3726,22 @@ def arange(
-        raise AssertionError()
+    assert not pin_memory
+    assert layout == torch.strided
+    # Case: torch.arange(5)
```
Nice comment

mruberry reviewed (approved):
Overall looks great. See inline comments. Also let's update the docs to clarify that the value is not checked here: https://pytorch.org/docs/master/special.html

lezcano committed:
Update on "Adding multigammaln ref and fix arange"


Partially based on https://github.com/pytorch/pytorch/pull/83662.

I'll help land this one, as Rob does not work in the PyTorch project
anymore

I removed the data-dependent check for the args, as data dependencies
are bad for many reasons (and it was failing when the input has NaNs).

It also registers arange as a decomposition, and fixes the naming of its
args.

[ghstack-poisoned]

Lezcano commented:
@pytorchbot merge -l

Lezcano commented:
@ngimel could I have an OK on this one?

lezcano committed:
Update on "Adding multigammaln ref and fix arange"


Partially based on https://github.com/pytorch/pytorch/pull/83662.

I'll help land this one, as Rob does not work in the PyTorch project
anymore

I removed the data-dependent check for the args, as data dependencies
are bad for many reasons (and it was failing when the input has NaNs).

It also registers arange as a decomposition, and fixes the naming of its
args.

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/UnaryOps.cpp:
```diff
@@ -723,8 +723,7 @@ constexpr double QUARTER = 0.25;
 static inline void mvlgamma_check(const Tensor& self, int64_t p) {
-  TORCH_CHECK((self > HALF * (p - 1)).all().item<bool>(),
-              "All elements must be greater than (p-1)/2");
+  TORCH_CHECK(self.scalar_type() != kBool, "The input tensor may not be a boolean tensor.");
```
this looks like a different check than what used to be here?

lezcano replied:
As per the OP:
>I removed the data-dependent check for the args, as data dependencies
are bad for many reasons (and it was failing when the input has NaNs).

The new check is because some test was failing for bools, so I simply disallowed bools for this function, as they don't make any sense anyway.

lezcano commented:
@pytorchbot merge
